src/main/java/com/epam/aidial/core/security/AccessService.java

```diff
@@ -11,16 +12,31 @@
 import com.epam.aidial.core.storage.ResourceDescription;
 import com.epam.aidial.core.util.ProxyUtil;
 import com.epam.aidial.core.util.UrlUtil;
+import com.google.common.collect.Sets;
 import io.vertx.core.json.JsonObject;
 
+import java.util.EnumSet;
+import java.util.HashMap;
+import java.util.HashSet;
 import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.BiFunction;
+
 
 public class AccessService {
 
     private final EncryptionService encryptionService;
     private final ShareService shareService;
     private final RuleService ruleService;
     private final List<Rule> adminRules;
+    private final List<PermissionRule> permissionRules = List.of(
```

Comment:
Nit: most of the use-cases related to chat-completion and own resources. I would rearrange checks to the following:
checkOwnResource - cheep
checkAutoShared - cheep
checkAppResourceAccess - cheep
checkSharedResource - mid
checkPublishedResource - expensive
checkReviewResource - cheep (rare case)

Reply:
Done.